transhumandesign/kag-base
# [changed] Egg breaks upon impact, makes a crack sound and can be used as a projectile

## Status

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.

## Description

[changed] Egg breaks on impact with walls and enemy players
[changed] Egg thrown against enemy players will stun them for 10 ticks
[changed] Egg gib count from 6 to 4
[added] Egg crack sound

This PR makes it so Egg will break if hitting the ground or a wall too hard. 
`if (this.getShape().vellen <= 6.8f) return;` in `onCollision()` is used which is equivalent to the check that causes the water splashing when a full Bucket hits the ground or a wall.

Egg can be thrown against an enemy to mini-stun them for 10 ticks.
This works if the Egg's velocity is higher than `6.8f`.
The Egg will break and not heal the enemy.
The enemy is set to dazzled so the screen flashes briefly which fits the effect of getting an Egg thrown at.
Although `this.server_Hit(blob, pos, Vec2f(), 0.0f, Hitters::water_stun_force, true);` could be used to create dazzle animation, screen flash, stun and ArgShort sound, it was not used because I cannot set the stun to 10 ticks and Sponge stun reduction might take effect unwantedly.

A crack sound has been added which plays upon hatching, hitting the ground or hitting an enemy.
It does not play when healing oneself or when Egg collides with Fireplace to create a Cake.

This has been tested offline and online and seems to work as intended.
The `isServer()` checks have been placed carefully to make sure the behavior is the same offline and online.

## Steps to Test or Reproduce

Go to offline or online Sandbox.
Throw an Egg down on the ground or run forward and throw it against a wall.
Notice Egg will break if hitting too hard.
Notice the Egg's crack sound.

Spawn an enemy builder, hit him once.
Run forward and throw an Egg against an enemy player.
Notice Egg will break and mini-stun the enemy player without healing him.
Notice the Egg's crack sound.

Place an Egg and wait for a chicken to hatch.
Notice the Egg's crack sound.

Place an Egg somewhere and hurt yourself, then pick up the Egg.
Notice the behavior is the same as before (unchanged).

Throw an Egg against a Fireplace.
Notice the behavior is the same as before (unchanged).

Place a Trampoline on a field with 2 empty spaces.
Have your Eggs bounce on it.
Change team to something else.
Go where the bouncing Eggs are.
Notice you will get mini-stunned.
